### ISSUE TITLE
Mackinaw ORE Development Edition

### ISSUE THREAD

**Issue by blitzmann** [DarkFenX/Pyfa#301]
I seem to have an old fit in the database that was never converted (probably happened when they released SKINs). Breaks fit backup as typeID 33683 cannot be found.

We should probably 1) add it to list of conversions, or 2) find a way to simply skip things that cannot be found (will also help with old modules that were missed and now fit crashes)

**Comment by blitzmann** [comment]
So, there may be a bit of an issue with this. There is no easy way that I can see to delete the fit if the ship does not exist, especially in this case where the fits are gathered at the DB level (queries.getAllFits or something). If we were using sFit.getFit() or similar we could add a flag to delete fit from service layer and check, but these fits are all loaded as one big query. And I'm not sure how, if it's possible, to delete the fit from inside eos.saveddata.ship itself (where the error pops up)

Anyway, this Mackinaw should be added to an upgrade. There was also a Rorqual ORE skin, not sure if it was ever a stand alone ship in pyfa. Gonna need to check previous versions later.

**Comment by DarkFenX** [comment]
Both of these were never as standalone ship.

We probably have to compose list of allowed items (taking overrides into consideration) during startup and remove anything which is not on that list (fit items - when loading a fit, ships - when loading a list with said ship). They might've been imported as EFT format before we added conversions, or maybe imported by other means which don't follow conversion rules (like DNA).

**Comment by blitzmann** [comment]
What do you think of this:

    try:
        self.ship = Ship(db.getItem(self.shipID)) if self.shipID is not None else None
    except:
        self.ship = None

This seems to work okay, and is fairly simple. If we set ship to None, then the export function works. There's an error when searching for "Mac" because it's trying to return some attributes that don't exist - should be easy to fix. Using this method, the object will initialize correctly, which is our primary concern. I think this would be okay for ships, and we may be able to use something similar for modules? I'll also look around and see if there's a way in SQLAlchemy to delete in the database

The Mackinaw still shows up in the xml export, but I think during an import it will be discarded.

Also, is DNA import really disregarded when it comes to conversions? We should probably fix that.

**Comment by DarkFenX** [comment]
I don't know how exactly it should be done. Generally speaking there're 2 approaches:

1) Ignoring corrupted module/ship in db
2) Removing corrupted module/ship from db

I am talking just about detection for now. For it to work, we have to export all the conversion data to some simple modules (which will not contain any complex market definitions or migration scripts) which can be exported from both eos' db layer and pyfa's market service, as well as from other places where needed (code which loads modules/fits, for example). Somewhere we should also place listing of all valid typeIDs - which are loaded from staticdata, taking overrides into consideration (overrides should also be placed into that simple module for this to work - if we want to have them taken into consideration on low level). This listing/set is then used during fit loading (or some periodic check process), and action is taken against offending typeIDs.

Regarding DNA - we keep mappings of names for imports, while DNA operates typeIDs. I don't think we convert these names to typeIDs (it's not possible for some of them at this stage), so i assume they were imported as-is - i.e. mackinaw ORE edition when it existed in database was imported, but not shown under mackinaw category.

**Comment by blitzmann** [comment]
My example was just a quick solution to avoid crashes, but yeah doesn't address the underlying issue. 

The problem here is that the object fails to initialize during a query call, which causes the query to fail. We could check if it's a valid item in the `init()` with the method you describe, but not sure where that will get us as we cannot take any action such as ignoring or deleting from inside the query call (that I have been able to find). We already check to make sure it's in the eve database with the call to `getItem()`, but we can't so anything about it.

I guess we could just move the ship and mode setting out of the `init()`. That would allow the object to initialize, and then when we try to fetch the ship information was can catch the exception. But, I just don't see an easy way to remove these outdated ships and modules in the database since there is no singular place where we can catch it, except at the `init()` which I haven't found a way to work yet. As an extension of that, I'm not sure where we can apply a list of valid typeIDs in situations where the item is loaded from a query instead of from the service or something similar.

An idea I've been thinking about ever since we started getting complaints about fits not loading is perhaps making a scrubbing function that simply loads all fits, along with their modules, characters, projected items, etc. If there's anything wrong with it, delete the fit and dump it to a log file to have the user sort it out. This way, if the user is having issues, they can start the database scrubber thing. Or, we can run it after every update or something. Considering that I think this is a relatively infrequent issue, I think this would work well. Unfortunately, I think some database corruption is inevitable over the long run just from the nature of the program, where we have the eve database that changes constantly.
